File: src/platform_context_graph/tools/languages/templated_detection.py
"""Dialect detection helpers for templated infrastructure files."""

from dataclasses import dataclass
from pathlib import Path
import re


GENERATED_DIRS = frozenset(
    {
        ".git",
        ".terraform",
        ".terragrunt-cache",
        ".venv",
        ".worktrees",
        "__pycache__",
        "build",
        "dist",
        "node_modules",
        "vendor",
    }
)
YAML_SUFFIXES = {".yaml", ".yml"}
HCL_SUFFIXES = {".hcl", ".tf", ".tfvars"}
JINJA_TEMPLATE_SUFFIXES = {".jinja", ".jinja2", ".j2"}
TERRAFORM_TEMPLATE_SUFFIXES = {".tpl", ".tftpl"}
RAW_CONFIG_SUFFIXES = {".conf", ".cfg", ".cnf"}
TEXT_SUFFIXES = YAML_SUFFIXES | HCL_SUFFIXES | JINJA_TEMPLATE_SUFFIXES | {
    ".kcl"
} | TERRAFORM_TEMPLATE_SUFFIXES | RAW_CONFIG_SUFFIXES
TEXT_FILENAMES = {
    "dockerfile",
    "compose.yaml",
    "compose.yml",
    "docker-compose.yaml",
    "docker-compose.yml",
}
GO_EXPRESSION_RE = re.compile(r"(?<!\$)\{\{[-~]?.*?[-~]?\}\}", re.DOTALL)
JINJA_STATEMENT_RE = re.compile(r"\{%-?.*?-?%\}|\{#.*?#\}", re.DOTALL)
GITHUB_ACTIONS_EXPR_RE = re.compile(r"\$\{\{.*?\}\}", re.DOTALL)
GO_CONTEXT_RE = re.compile(r"\{\{[-~]?\s*(?:\.|\$)")
GO_LINE_CONTROL_RE = re.compile(
    r"(?m)^\s*\{\{[-~]?\s*(if|else|end|with|range|define|template|block)\b"
)
GO_HINT_RE = re.compile(r"\b(include|toYaml|nindent|tpl)\b")
TF_INTERPOLATION_RE = re.compile(r"(?<!\$)\$\{")
TF_DIRECTIVE_RE = re.compile(r"(?<!%)%\{")
TF_TEMPLATEFILE_RE = re.compile(r"\btemplatefile\s*\(")


@dataclass(frozen=True, slots=True)
class FileClassification:
    """Describe one authored file classified by dialect heuristics."""

    relative_path: Path
    bucket: str
    dialects: tuple[str, ...]
    ambiguous: bool
    marker_count: int
    marker_density: float
    renderability_hint: str
    artifact_type: str
    raw_ingest_candidate: bool
    iac_relevant: bool


@dataclass(frozen=True, slots=True)
class ContentMetadata:
    """Persisted metadata attached to indexed file and entity content rows."""

    artifact_type: str | None
    template_dialect: str | None
    iac_relevant: bool


def exclusion_reason(relative_path: Path, *, include_generated: bool) -> str | None:
    """Return the generated-directory reason for excluding a file, if any."""

    if include_generated:
        return None
    for part in relative_path.parts:
        if part in GENERATED_DIRS:
            return part
    return None


def infer_content_metadata(*, relative_path: Path, content: str) -> ContentMetadata:
    """Infer persisted content metadata without requiring inventory-only context."""

    if not is_candidate_text_file(relative_path):
        return ContentMetadata(
            artifact_type=None,
            template_dialect=None,
            iac_relevant=False,
        )

    classification = classify_file(
        root_family=_infer_root_family(relative_path, content),
        relative_path=relative_path,
        content=content,
    )
    return ContentMetadata(
        artifact_type=_persisted_artifact_type(classification),
        template_dialect=_persisted_template_dialect(classification),
        iac_relevant=classification.iac_relevant,
    )


def is_candidate_text_file(path: Path) -> bool:
    """Return whether the path should be included in the inventory scan."""

    name = path.name.lower()
    return name in TEXT_FILENAMES or path.suffix.lower() in TEXT_SUFFIXES


def _infer_root_family(relative_path: Path, content: str) -> str:
    """Infer a production classification family from path and content hints."""

    parts = {part.lower() for part in relative_path.parts}
    name = relative_path.name.lower()
    suffixes = _suffixes(relative_path)
    go_expression_match = GO_EXPRESSION_RE.search(content)
    has_tf_markers = bool(
        TF_INTERPOLATION_RE.search(content)
        or TF_DIRECTIVE_RE.search(content)
        or TF_TEMPLATEFILE_RE.search(content)
    )
    if any(suffix in HCL_SUFFIXES for suffix in suffixes):
        return "terraform"
    if has_tf_markers and any(
        suffix in TERRAFORM_TEMPLATE_SUFFIXES | JINJA_TEMPLATE_SUFFIXES
        for suffix in suffixes
    ) and not (go_expression_match or JINJA_STATEMENT_RE.search(content)):
        return "terraform"
    if (
        suffixes
        and suffixes[-1] == ".tpl"
        and "templates" in parts
        and go_expression_match
        and (
            name == "_helpers.tpl"
            or any(
                marker in content
                for marker in (
                    ".Chart",
                    ".Release",
                    ".Values",
                    '{{ include "',
                    '{{- include "',
                    '{{ define "',
                    '{{- define "',
                )
            )
        )
    ):
        return "helm_argo"
    if (
        name == "chart.yaml"
        or name.startswith("values.")
        or ("chart" in parts and "templates" in parts)
        or "argocd" in parts
    ):
        return "helm_argo"
    if {
        "roles",
        "playbooks",
        "handlers",
        "tasks",
        "group_vars",
        "host_vars",
        "inventory",
        "inventories",
    } & parts:
        return "ansible_jinja"
    if {"dagster", "assets", "data_quality", "data_lakehouse"} & parts:
        return "dagster_jinja"
    return "generic"


def _suffixes(relative_path: Path) -> tuple[str, ...]:
    """Return normalized suffixes for a path."""

    return tuple(suffix.lower() for suffix in relative_path.suffixes)


def _artifact_type(root_family: str, relative_path: Path) -> str:
    """Infer a coarse artifact type for reporting."""

    name = relative_path.name.lower()
    parts = {part.lower() for part in relative_path.parts}
    suffixes = _suffixes(relative_path)
    is_template_suffix = bool(
        suffixes
        and suffixes[-1] in JINJA_TEMPLATE_SUFFIXES | TERRAFORM_TEMPLATE_SUFFIXES
    )

    if ".github" in parts and "workflows" in parts:
        return "github_actions_workflow"
    if name == "dockerfile" or name.startswith("dockerfile."):
        return "dockerfile"
    if name in {"compose.yaml", "compose.yml", "docker-compose.yaml", "docker-compose.yml"}:
        return "docker_compose"
    if RAW_CONFIG_SUFFIXES.intersection(suffixes):
        if "apache" in parts or "httpd" in parts or "mods-available" in parts:
            return "apache_config_template" if is_template_suffix else "apache_config"
        if "nginx" in parts:
            return "nginx_config_template" if is_template_suffix else "nginx_config"
        return "generic_config_template" if is_template_suffix else "generic_config"
    if any(suffix in YAML_SUFFIXES for suffix in suffixes) and any(
        suffix in JINJA_TEMPLATE_SUFFIXES for suffix in suffixes
    ):
        return "yaml_template"
    if suffixes and suffixes[-1] in JINJA_TEMPLATE_SUFFIXES:
        if root_family == "terraform":
            return "terraform_template_text"
        return "jinja_text_template"
    if suffixes and suffixes[-1] in TERRAFORM_TEMPLATE_SUFFIXES:
        return "terraform_template_text" if root_family == "terraform" else "text_template"
    if any(suffix in HCL_SUFFIXES for suffix in suffixes):
        return "terraform_hcl"
    if any(suffix in YAML_SUFFIXES for suffix in suffixes):
        return "yaml_document"
    return "plain_text"


def _persisted_artifact_type(classification: FileClassification) -> str | None:
    """Map inventory classification buckets to persisted content metadata."""

    if classification.bucket == "helm_helper_tpl":
        return "helm_helper_tpl"
    if classification.bucket == "go_template_yaml":
        return "go_template_yaml"
    if classification.bucket == "jinja_yaml":
        return "jinja_yaml"
    if classification.bucket in {"terraform_hcl", "terraform_hcl_templated"}:
        return "terraform_hcl"
    if classification.artifact_type in {"plain_text", "yaml_document"}:
        return (
            classification.artifact_type
            if classification.iac_relevant or classification.dialects
            else None
        )
    return classification.artifact_type


def _persisted_template_dialect(classification: FileClassification) -> str | None:
    """Normalize detected dialects for persisted content metadata."""

    if classification.ambiguous or not classification.dialects:
        return None
    dialect = classification.dialects[0]
    if dialect == "jinja_template":
        return "jinja"
    return dialect


def _is_raw_ingest_candidate(*, artifact_type: str, bucket: str) -> bool:
    """Return whether the file is a raw-text indexing gap today."""

    return artifact_type in {
        "apache_config",
        "apache_config_template",
        "dockerfile",
        "generic_config_template",
        "generic_config",
        "jinja_text_template",
        "nginx_config",
        "nginx_config_template",
        "terraform_template_text",
        "text_template",
    } or bucket == "plain_text"


def _is_iac_relevant(
    *,
    root_family: str,
    relative_path: Path,
    artifact_type: str,
    bucket: str,
) -> bool:
    """Return whether the file is IaC-relevant for reporting."""

    if artifact_type == "github_actions_workflow":
        return False
    if root_family in {"helm_argo", "ansible_jinja", "terraform"}:
        return True
    if artifact_type in {
        "apache_config",
        "apache_config_template",
        "docker_compose",
        "dockerfile",
        "nginx_config",
        "nginx_config_template",
        "terraform_template_text",
        "yaml_template",
        "generic_config_template",
    }:
        return True
    if bucket in {
        "go_template_yaml",
        "jinja_yaml",
        "terraform_hcl",
        "terraform_hcl_templated",
    }:
        return True
    return "iac" in {part.lower() for part in relative_path.parts}


def classify_file(
    *,
    root_family: str,
    relative_path: Path,
    content: str,
) -> FileClassification:
    """Classify one file by path and templating heuristics."""

    suffix = relative_path.suffix.lower()
    suffixes = _suffixes(relative_path)
    artifact_type = _artifact_type(root_family, relative_path)
    lowered_content = content.lower()
    if artifact_type in {"generic_config", "generic_config_template"}:
        is_template_config = artifact_type.endswith("_template")
        if any(
            token in lowered_content
            for token in ("server {", "fastcgi_pass", "proxy_pass", "location /")
        ):
            artifact_type = (
                "nginx_config_template" if is_template_config else "nginx_config"
            )
        elif any(
            token in lowered_content
            for token in (
                "<virtualhost",
                "rewriterule",
                "documentroot",
                "servername ",
            )
        ):
            artifact_type = (
                "apache_config_template" if is_template_config else "apache_config"
            )
    yaml_template = (
        len(suffixes) >= 2
        and suffixes[-1] in JINJA_TEMPLATE_SUFFIXES
        and suffixes[-2] in YAML_SUFFIXES
    )
    line_count = max(content.count("\n") + 1, 1)
    go_expressions = GO_EXPRESSION_RE.findall(content)
    go_expression_count = len(go_expressions)
    jinja_statement_count = len(JINJA_STATEMENT_RE.findall(content))
    github_actions_count = len(GITHUB_ACTIONS_EXPR_RE.findall(content))
    go_context_count = len(GO_CONTEXT_RE.findall(content))
    go_line_control_count = len(GO_LINE_CONTROL_RE.findall(content))
    go_hint_count = sum(len(GO_HINT_RE.findall(expr)) for expr in go_expressions)
    tf_interpolation_count = len(TF_INTERPOLATION_RE.findall(content))
    tf_directive_count = len(TF_DIRECTIVE_RE.findall(content))
    tf_templatefile_count = len(TF_TEMPLATEFILE_RE.findall(content))
    tf_marker_count = (
        tf_interpolation_count + tf_directive_count + tf_templatefile_count
    )
    marker_count = (
        go_expression_count
        + jinja_statement_count
        + github_actions_count
        + tf_marker_count
    )
    marker_density = marker_count / line_count

    def build_result(
        *,
        bucket: str,
        dialects: tuple[str, ...],
        ambiguous: bool,
        renderability_hint: str,
        count: int = marker_count,
        density: float = marker_density,
    ) -> FileClassification:
        """Create one classification result with reporting metadata applied."""

        return FileClassification(
            relative_path=relative_path,
            bucket=bucket,
            dialects=dialects,
            ambiguous=ambiguous,
            marker_count=count,
            marker_density=density,
            renderability_hint=renderability_hint,
            artifact_type=artifact_type,
            raw_ingest_candidate=_is_raw_ingest_candidate(
                artifact_type=artifact_type,
                bucket=bucket,
            ),
            iac_relevant=_is_iac_relevant(
                root_family=root_family,
                relative_path=relative_path,
                artifact_type=artifact_type,
                bucket=bucket,
            ),
        )

    if suffix == ".tpl" and root_family == "helm_argo":
        return build_result(
            bucket="helm_helper_tpl",
            dialects=("go_template",),
            ambiguous=False,
            renderability_hint="context_required",
        )

    if yaml_template:
        suffix = suffixes[-2]

    if suffix in JINJA_TEMPLATE_SUFFIXES and root_family == "terraform":
        template_count = max(tf_marker_count, marker_count)
        return build_result(
            bucket="unknown_templated",
            dialects=("terraform_template",),
            ambiguous=False,
            renderability_hint="context_required",
            count=template_count,
            density=template_count / line_count,
        )

    if suffix in JINJA_TEMPLATE_SUFFIXES:
        return build_result(
            bucket="unknown_templated",
            dialects=("jinja_template",),
            ambiguous=False,
            renderability_hint="context_required",
        )

    if suffix in TERRAFORM_TEMPLATE_SUFFIXES and root_family == "terraform":
        template_count = max(tf_marker_count, marker_count)
        return build_result(
            bucket="unknown_templated",
            dialects=("terraform_template",),
            ambiguous=False,
            renderability_hint="context_required",
            count=template_count,
            density=template_count / line_count,
        )

    if suffix in HCL_SUFFIXES:
        bucket = "terraform_hcl_templated" if tf_marker_count else "terraform_hcl"
        dialects = ("terraform_template",) if tf_marker_count else ()
        renderability = "context_required" if tf_marker_count else "raw_only"
        return build_result(
            bucket=bucket,
            dialects=dialects,
            ambiguous=False,
            renderability_hint=renderability,
            count=tf_marker_count,
            density=tf_marker_count / line_count,
        )

    if suffix not in YAML_SUFFIXES and suffix != ".kcl":
        return build_result(
            bucket="unknown_templated" if marker_count else "plain_text",
            dialects=(),
            ambiguous=False,
            renderability_hint="context_required" if marker_count else "raw_only",
        )

    explicit_go = bool(go_line_control_count or go_hint_count or go_context_count)
    explicit_jinja = bool(jinja_statement_count)
    has_curly_expressions = bool(go_expression_count)
    has_github_actions = bool(github_actions_count)

    if has_github_actions and not (explicit_go or explicit_jinja or has_curly_expressions):
        return build_result(
            bucket="unknown_templated",
            dialects=("github_actions",),
            ambiguous=False,
            renderability_hint="context_required",
        )

    if not (explicit_go or explicit_jinja or has_curly_expressions):
        return build_result(
            bucket="plain_yaml",
            dialects=(),
            ambiguous=False,
            renderability_hint="raw_only",
            count=0,
            density=0.0,
        )

    if explicit_go and explicit_jinja:
        return build_result(
            bucket="unknown_templated",
            dialects=("go_template", "jinja"),
            ambiguous=True,
            renderability_hint="raw_only",
        )

    if root_family in {"ansible_jinja", "dagster_jinja"}:
        if explicit_go:
            return build_result(
                bucket="unknown_templated",
                dialects=("go_template", "jinja"),
                ambiguous=True,
                renderability_hint="raw_only",
            )
        return build_result(
            bucket="jinja_yaml",
            dialects=("jinja",),
            ambiguous=False,
            renderability_hint="context_required",
        )

    if root_family == "helm_argo":
        if explicit_jinja:
            return build_result(
                bucket="unknown_templated",
                dialects=("go_template", "jinja"),
                ambiguous=True,
                renderability_hint="raw_only",
            )
        return build_result(
            bucket="go_template_yaml",
            dialects=("go_template",),
            ambiguous=False,
            renderability_hint="context_required",
        )

    if explicit_jinja and has_curly_expressions:
        dialects = ("go_template", "jinja")
    elif explicit_jinja:
        dialects = ("jinja",)
    elif explicit_go:
        dialects = ("go_template",)
    elif has_curly_expressions:
        dialects = ("go_template", "jinja")
    else:
        dialects = ()

    if len(dialects) == 1:
        bucket = "go_template_yaml" if dialects[0] == "go_template" else "jinja_yaml"
        ambiguous = False
        renderability_hint = "context_required"
    else:
        bucket = "unknown_templated"
        ambiguous = True
        renderability_hint = "raw_only"

    return build_result(
        bucket=bucket,
        dialects=dialects,
        ambiguous=ambiguous,
        renderability_hint=renderability_hint,
    )

File: go/internal/parser/templated_detection.go
package parser

import (
	"path/filepath"
	"regexp"
	"strings"
)

type contentMetadata struct {
	ArtifactType    string
	TemplateDialect string
	IACRelevant     bool
}

var (
	goExpressionRE      = regexp.MustCompile(`(?s)\{\{[-~]?.*?[-~]?\}\}`)
	jinjaStatementRE    = regexp.MustCompile(`(?s)\{%-?.*?-?%\}|\{#.*?#\}`)
	githubActionsExprRE = regexp.MustCompile(`(?s)\$\{\{.*?\}\}`)
	goContextRE         = regexp.MustCompile(`\{\{[-~]?\s*(?:\.|\$)`)
	goLineControlRE     = regexp.MustCompile(`(?m)^\s*\{\{[-~]?\s*(if|else|end|with|range|define|template|block)\b`)
	goHintRE            = regexp.MustCompile(`\b(include|toYaml|nindent|tpl)\b`)
	tfInterpolationRE   = regexp.MustCompile(`\$\{`)
	tfDirectiveRE       = regexp.MustCompile(`%\{`)
	tfTemplatefileRE    = regexp.MustCompile(`\btemplatefile\s*\(`)
)

func inferContentMetadata(path string, content string) contentMetadata {
	rootFamily := inferRootFamily(path, content)
	artifactType := inferArtifactType(rootFamily, path, content)
	loweredContent := strings.ToLower(content)
	if artifactType == "generic_config" || artifactType == "generic_config_template" {
		switch {
		case strings.Contains(loweredContent, "server {") ||
			strings.Contains(loweredContent, "fastcgi_pass") ||
			strings.Contains(loweredContent, "proxy_pass") ||
			strings.Contains(loweredContent, "location /"):
			if strings.HasSuffix(artifactType, "_template") {
				artifactType = "nginx_config_template"
			} else {
				artifactType = "nginx_config"
			}
		case strings.Contains(loweredContent, "<virtualhost") ||
			strings.Contains(loweredContent, "rewriterule") ||
			strings.Contains(loweredContent, "documentroot") ||
			strings.Contains(loweredContent, "servername "):
			if strings.HasSuffix(artifactType, "_template") {
				artifactType = "apache_config_template"
			} else {
				artifactType = "apache_config"
			}
		}
	}

	suffixes := splitSuffixes(path)
	lastSuffix := ""
	if len(suffixes) > 0 {
		lastSuffix = suffixes[len(suffixes)-1]
	}
	if len(suffixes) >= 2 && isJinjaSuffix(suffixes[len(suffixes)-1]) && isYAMLSuffix(suffixes[len(suffixes)-2]) {
		lastSuffix = suffixes[len(suffixes)-2]
	}

	goExpressions := filteredMatches(content, goExpressionRE, '$')
	explicitGo := goContextRE.MatchString(content) || goLineControlRE.MatchString(content)
	for _, expression := range goExpressions {
		if len(goHintRE.FindAllString(expression, -1)) > 0 {
			explicitGo = true
			break
		}
	}
	explicitJinja := jinjaStatementRE.MatchString(content)
	hasCurlyExpressions := len(goExpressions) > 0
	hasGitHubActions := githubActionsExprRE.MatchString(content)
	tfMarkerCount := countUnprefixedMatches(content, tfInterpolationRE, '$') +
		countUnprefixedMatches(content, tfDirectiveRE, '%') +
		len(tfTemplatefileRE.FindAllString(content, -1))

	templateDialect := ""
	bucket := "plain_text"
	switch {
	case lastSuffix == ".tpl" && rootFamily == "helm_argo":
		bucket = "helm_helper_tpl"
		templateDialect = "go_template"
	case isJinjaSuffix(lastSuffix) && rootFamily == "terraform":
		bucket = "unknown_templated"
		templateDialect = "terraform_template"
	case isJinjaSuffix(lastSuffix):
		bucket = "unknown_templated"
		templateDialect = "jinja"
	case isTerraformTemplateSuffix(lastSuffix) && rootFamily == "terraform":
		bucket = "unknown_templated"
		templateDialect = "terraform_template"
	case isHCLSuffix(lastSuffix):
		bucket = "terraform_hcl"
		if tfMarkerCount > 0 {
			bucket = "terraform_hcl_templated"
			templateDialect = "terraform_template"
		}
	case strings.HasPrefix(artifactType, "ansible_"):
		bucket = artifactType
		if explicitJinja || jinjaStatementRE.MatchString(content) || hasCurlyExpressions {
			templateDialect = "jinja"
		}
	case !isYAMLSuffix(lastSuffix) && lastSuffix != ".kcl":
		if templateDialect == "" && (explicitJinja || hasCurlyExpressions || hasGitHubActions || tfMarkerCount > 0) {
			bucket = "unknown_templated"
		}
	case hasGitHubActions && !explicitGo && !explicitJinja && !hasCurlyExpressions:
		bucket = "unknown_templated"
		templateDialect = "github_actions"
	case !explicitGo && !explicitJinja && !hasCurlyExpressions:
		bucket = "plain_yaml"
	case explicitGo && explicitJinja:
		bucket = "unknown_templated"
	case rootFamily == "ansible_jinja" || rootFamily == "dagster_jinja":
		if explicitGo {
			bucket = "unknown_templated"
		} else {
			bucket = "jinja_yaml"
			templateDialect = "jinja"
		}
	case rootFamily == "helm_argo":
		if explicitJinja {
			bucket = "unknown_templated"
		} else {
			bucket = "go_template_yaml"
			templateDialect = "go_template"
		}
	case explicitJinja && hasCurlyExpressions:
		bucket = "unknown_templated"
	case explicitJinja:
		bucket = "jinja_yaml"
		templateDialect = "jinja"
	case explicitGo:
		bucket = "go_template_yaml"
		templateDialect = "go_template"
	case hasCurlyExpressions:
		bucket = "unknown_templated"
	}

	return contentMetadata{
		ArtifactType:    persistedArtifactType(bucket, artifactType),
		TemplateDialect: templateDialect,
		IACRelevant:     isIACRelevant(rootFamily, path, artifactType, bucket),
	}
}

func filteredMatches(content string, expression *regexp.Regexp, disallowedPrefix byte) []string {
	indexes := expression.FindAllStringIndex(content, -1)
	matches := make([]string, 0, len(indexes))
	for _, index := range indexes {
		start := index[0]
		if start > 0 && content[start-1] == disallowedPrefix {
			continue
		}
		matches = append(matches, content[index[0]:index[1]])
	}
	return matches
}

func countUnprefixedMatches(content string, expression *regexp.Regexp, disallowedPrefix byte) int {
	return len(filteredMatches(content, expression, disallowedPrefix))
}

func inferRootFamily(path string, content string) string {
	parts := pathParts(path)
	name := strings.ToLower(filepath.Base(path))
	suffixes := splitSuffixes(path)
	hasTFMarkers := tfInterpolationRE.MatchString(content) ||
		tfDirectiveRE.MatchString(content) ||
		tfTemplatefileRE.MatchString(content)

	switch {
	case anySuffix(suffixes, isHCLSuffix):
		return "terraform"
	case hasTFMarkers &&
		(anySuffix(suffixes, isTerraformTemplateSuffix) || anySuffix(suffixes, isJinjaSuffix)) &&
		!goExpressionRE.MatchString(content) &&
		!jinjaStatementRE.MatchString(content):
		return "terraform"
	case len(suffixes) > 0 && suffixes[len(suffixes)-1] == ".tpl" &&
		hasPart(parts, "templates") && goExpressionRE.MatchString(content) &&
		(name == "_helpers.tpl" || strings.Contains(content, ".Chart") ||
			strings.Contains(content, ".Release") || strings.Contains(content, ".Values") ||
			strings.Contains(content, `{{ include "`) || strings.Contains(content, `{{- include "`) ||
			strings.Contains(content, `{{ define "`) || strings.Contains(content, `{{- define "`)):
		return "helm_argo"
	case name == "chart.yaml" || strings.HasPrefix(name, "values.") ||
		(hasPart(parts, "chart") && hasPart(parts, "templates")) ||
		hasPart(parts, "argocd"):
		return "helm_argo"
	case hasPart(parts, "roles", "playbooks", "handlers", "tasks", "group_vars", "host_vars", "inventory", "inventories"):
		return "ansible_jinja"
	case hasPart(parts, "dagster", "assets", "data_quality", "data_lakehouse"):
		return "dagster_jinja"
	default:
		return "generic"
	}
}

func inferArtifactType(rootFamily string, path string, content string) string {
	name := strings.ToLower(filepath.Base(path))
	parts := pathParts(path)
	suffixes := splitSuffixes(path)
	lastSuffix := ""
	if len(suffixes) > 0 {
		lastSuffix = suffixes[len(suffixes)-1]
	}
	isTemplateSuffix := isJinjaSuffix(lastSuffix) || isTerraformTemplateSuffix(lastSuffix)
	ansibleType := ansibleArtifactType(parts, name, content)

	switch {
	case ansibleType != "":
		return ansibleType
	case hasPart(parts, ".github") && hasPart(parts, "workflows"):
		return "github_actions_workflow"
	case name == "dockerfile" || strings.HasPrefix(name, "dockerfile."):
		return "dockerfile"
	case isDockerComposeFilename(name):
		return "docker_compose"
	case anySuffix(suffixes, isRawConfigSuffix):
		switch {
		case hasPart(parts, "apache", "httpd", "mods-available"):
			if isTemplateSuffix {
				return "apache_config_template"
			}
			return "apache_config"
		case hasPart(parts, "nginx"):
			if isTemplateSuffix {
				return "nginx_config_template"
			}
			return "nginx_config"
		default:
			if isTemplateSuffix {
				return "generic_config_template"
			}
			return "generic_config"
		}
	case anySuffix(suffixes, isYAMLSuffix) && anySuffix(suffixes, isJinjaSuffix):
		return "yaml_template"
	case isJinjaSuffix(lastSuffix):
		if rootFamily == "terraform" {
			return "terraform_template_text"
		}
		return "jinja_text_template"
	case isTerraformTemplateSuffix(lastSuffix):
		if rootFamily == "terraform" {
			return "terraform_template_text"
		}
		return "text_template"
	case anySuffix(suffixes, isHCLSuffix):
		return "terraform_hcl"
	case anySuffix(suffixes, isYAMLSuffix):
		return "yaml_document"
	default:
		return "plain_text"
	}
}

func ansibleArtifactType(parts map[string]struct{}, name string, content string) string {
	switch {
	case hasPart(parts, "inventories", "inventory"):
		return "ansible_inventory"
	case hasPart(parts, "group_vars", "host_vars"):
		return "ansible_vars"
	case hasPart(parts, "playbooks"):
		return "ansible_playbook"
	case hasPart(parts, "roles"):
		if hasPart(parts, "tasks") && (name == "main.yml" || name == "main.yaml") {
			return "ansible_task_entrypoint"
		}
		if hasPart(parts, "vars", "defaults") {
			return "ansible_vars"
		}
		return "ansible_role"
	case isAnsiblePlaybookContent(content):
		return "ansible_playbook"
	default:
		return ""
	}
}

func isAnsiblePlaybookContent(content string) bool {
	lowered := strings.ToLower(content)
	return strings.Contains(lowered, "\nhosts:") ||
		strings.Contains(lowered, "\nroles:") ||
		strings.Contains(lowered, "\nvars_files:") ||
		strings.Contains(lowered, "\nimport_playbook:")
}

func persistedArtifactType(bucket string, artifactType string) string {
	switch bucket {
	case "helm_helper_tpl":
		return "helm_helper_tpl"
	case "go_template_yaml":
		return "go_template_yaml"
	case "jinja_yaml":
		return "jinja_yaml"
	case "terraform_hcl", "terraform_hcl_templated":
		return "terraform_hcl"
	default:
		if artifactType == "plain_text" || artifactType == "yaml_document" {
			return ""
		}
		return artifactType
	}
}

func isIACRelevant(rootFamily string, path string, artifactType string, bucket string) bool {
	if artifactType == "github_actions_workflow" {
		return false
	}
	switch rootFamily {
	case "helm_argo", "ansible_jinja", "terraform", "dagster_jinja":
		return true
	}
	switch artifactType {
	case "apache_config", "apache_config_template", "docker_compose", "dockerfile",
		"nginx_config", "nginx_config_template", "terraform_template_text",
		"yaml_template", "generic_config_template":
		return true
	}
	switch bucket {
	case "go_template_yaml", "jinja_yaml", "terraform_hcl", "terraform_hcl_templated":
		return true
	default:
		return hasPart(pathParts(path), "iac")
	}
}

func pathParts(path string) map[string]struct{} {
	parts := make(map[string]struct{})
	for _, part := range strings.Split(filepath.ToSlash(path), "/") {
		normalized := strings.ToLower(strings.TrimSpace(part))
		if normalized != "" {
			parts[normalized] = struct{}{}
		}
	}
	return parts
}

func hasPart(parts map[string]struct{}, values ...string) bool {
	for _, value := range values {
		if _, ok := parts[strings.ToLower(value)]; ok {
			return true
		}
	}
	return false
}

func anySuffix(suffixes []string, match func(string) bool) bool {
	for _, suffix := range suffixes {
		if match(suffix) {
			return true
		}
	}
	return false
}

func isYAMLSuffix(suffix string) bool {
	return suffix == ".yaml" || suffix == ".yml"
}

func isHCLSuffix(suffix string) bool {
	return suffix == ".hcl" || suffix == ".tf" || suffix == ".tfvars"
}

func isJinjaSuffix(suffix string) bool {
	return suffix == ".jinja" || suffix == ".jinja2" || suffix == ".j2"
}

func isTerraformTemplateSuffix(suffix string) bool {
	return suffix == ".tpl" || suffix == ".tftpl"
}

func isRawConfigSuffix(suffix string) bool {
	return suffix == ".conf" || suffix == ".cfg" || suffix == ".cnf"
}

func isDockerComposeFilename(name string) bool {
	return name == "compose.yaml" ||
		name == "compose.yml" ||
		name == "docker-compose.yaml" ||
		name == "docker-compose.yml" ||
		(strings.HasPrefix(name, "docker-compose.") && (strings.HasSuffix(name, ".yaml") || strings.HasSuffix(name, ".yml")))
}
